File: src/main/java/io/github/kuroka3/playermanager/Commands/TempBan.java
package io.github.kuroka3.playermanager.Commands;

import io.github.kuroka3.playermanager.Class.ManagedPlayer;
import io.github.kuroka3.playermanager.PlayerManager;
import io.github.kuroka3.playermanager.Utils.BanIDManager;
import io.github.kuroka3.playermanager.Utils.CaseManager;
import io.github.kuroka3.playermanager.Utils.JSONFile;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.OfflinePlayer;
import org.bukkit.block.CommandBlock;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.command.ConsoleCommandSender;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

public class TempBan implements CommandExecutor {

    private static String mod = ChatColor.GRAY + "[" + ChatColor.LIGHT_PURPLE + "!" + ChatColor.GRAY + "] " + ChatColor.RESET;

    @Override
    public boolean onCommand(@NotNull CommandSender sd, @NotNull Command cmd, @NotNull String label, @NotNull String[] args) {
        try {
            if (sd instanceof CommandBlock) {
                return true;
            }

            if (args.length == 0) {
                sd.sendMessage(mod + ChatColor.RED + "플레이어를 지정해 주십시오");
                return true;
            }

            if (args.length < 3 || !(args[2].equalsIgnoreCase("seconds") || args[2].equalsIgnoreCase("minutes") || args[2].equalsIgnoreCase("hours") || args[2].equalsIgnoreCase("days") || args[2].equalsIgnoreCase("months") || args[2].equalsIgnoreCase("years"))) {
                sd.sendMessage(mod + ChatColor.RED + "사용법이 알맞지 않습니다");
                return true;
            }

            if (!sd.hasPermission("playermanager.ban") && !(sd instanceof ConsoleCommandSender)) {
                sd.sendMessage(mod + ChatColor.RED + "이 명령어를 사용할 권한이 없습니다");
                return true;
            }

            JSONFile playerJson = new JSONFile(PlayerManager.getPlugin(PlayerManager.class).getDataFolder() + "/players.json");

            Player player2 = Bukkit.getServer().getPlayer(args[0]);

            ManagedPlayer target;

            if(player2 == null) {
                OfflinePlayer offlinePlayer = Bukkit.getServer().getOfflinePlayer(args[0]);
                if(offlinePlayer.hasPlayedBefore()) {
                    target = new ManagedPlayer(offlinePlayer, playerJson);
                } else {
                    sd.sendMessage(mod + ChatColor.RED + "해당 유저를 찾을 수 없습니다.");
                    return true;
                }
            } else {
                target = new ManagedPlayer(player2, playerJson);
            }

            if (target.isBan()) {
                sd.sendMessage(mod + ChatColor.RED + "해당 유저는 이미 밴 상태입니다");
                return true;
            }

            StringBuilder sb = new StringBuilder();

            for (int i = 3; i < args.length; i++) {
                sb.append(args[i]);
                if (i != args.length - 1) {
                    sb.append(" ");
                }
            }

            String reason = sb.toString();

            String id = String.valueOf(
                    Long.parseLong(
                            LocalDateTime.now().format(
                                    DateTimeFormatter.ofPattern("yyyyMMddhhmmssSS")
                            )
                    )*5
            );

            LocalDateTime toban = LocalDateTime.now();

            if(args[2].equalsIgnoreCase("seconds")) toban = toban.plusSeconds(Long.parseLong(args[1]));
            if(args[2].equalsIgnoreCase("minutes")) toban = toban.plusMinutes(Long.parseLong(args[1]));
            if(args[2].equalsIgnoreCase("hours")) toban = toban.plusHours(Long.parseLong(args[1]));
            if(args[2].equalsIgnoreCase("days")) toban = toban.plusDays(Long.parseLong(args[1]));
            if(args[2].equalsIgnoreCase("months")) toban = toban.plusMonths(Long.parseLong(args[1]));
            if(args[2].equalsIgnoreCase("years")) toban = toban.plusYears(Long.parseLong(args[1]));

            target.tempban(reason, id, toban);

            String moder = null;

            if(sd instanceof Player player) {
                moder = player.getUniqueId().toString();
            } else if (sd instanceof ConsoleCommandSender) {
                moder = "Server";
            }

            String saveReason;

            if(reason.isEmpty()) {
                saveReason = "No Reason Given";
            } else {
                saveReason = reason;
            }

            if(!target.isOffline()) {
                BanIDManager.setBan(id, target.getPlayer().getUniqueId(), moder, saveReason, LocalDateTime.now());
                CaseManager.addCase(4, moder, target.getPlayer().getUniqueId(), saveReason, args[1] + args[2]);


                sd.sendMessage(mod + ChatColor.YELLOW + target.getPlayer().getName() + ChatColor.RED + "님을 서버에서 차단했습니다: " + ChatColor.GOLD + saveReason);
            } else {
                BanIDManager.setBan(id, target.getOfflinePlayer().getUniqueId(), moder, saveReason, LocalDateTime.now());
                CaseManager.addCase(4, moder, target.getOfflinePlayer().getUniqueId(), saveReason, args[1] + args[2]);


                sd.sendMessage(mod + ChatColor.YELLOW + target.getOfflinePlayer().getName() + ChatColor.RED + "님을 서버에서 차단했습니다: " + ChatColor.GOLD + saveReason);
            }



            return true;
        } catch (Exception e) {
            e.printStackTrace();
            sd.sendMessage(mod + ChatColor.RED + "Check Console: An Exception occured");
            return true;
        }
    }
}

File: src/main/kotlin/io/github/kuroka3/playermanager/Commands/TempBan.kt
package io.github.kuroka3.playermanager.Commands

import io.github.kuroka3.playermanager.Class.ManagedPlayer
import io.github.kuroka3.playermanager.PlayerManager
import io.github.kuroka3.playermanager.Utils.BanIDManager
import io.github.kuroka3.playermanager.Utils.CaseManager
import io.github.monun.kommand.StringType
import io.github.monun.kommand.getValue
import io.github.monun.kommand.kommand
import net.kyori.adventure.text.Component.text
import net.kyori.adventure.text.TextComponent
import net.kyori.adventure.text.format.TextColor.color
import net.kyori.adventure.title.Title
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import java.lang.Exception
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

object TempBan {
    fun registerKommand() {

        val PlayerManager = PlayerManager.instance

        val mod: TextComponent = text("[").color(color(0x00aaaaaa)).append(
            text("!").color(color(0x00ff55ff))
        ).append(
            text("] ").color(color(0x00aaaaaa))
        )

        PlayerManager.kommand {
            register("tempban") {
                requires { (isPlayer && hasPermission("playermanager.ban")) || isConsole }

                val onlinePlayer = string().apply {
                    suggests {
                        Bukkit.getOnlinePlayers().forEach {
                            suggest(it.name)
                        }
                    }
                }

                val reasonString = string(StringType.GREEDY_PHRASE).apply {
                    suggests {
                        suggest(
                            listOf(
                                "No Reason Given",
                                "Reason"
                            )
                        )
                    }
                }

                val periodSuggest = long().apply {
                    suggests {
                        suggest(
                            listOf(
                                "1",
                                "5",
                                "10",
                                "30",
                                "90",
                                "365"
                            )
                        )
                    }
                }

                val unitSuggest = string().apply {
                    suggests {
                        suggest(
                            listOf(
                                "days", "hours", "minutes", "seconds"
                            )
                        )
                    }
                }

                then("target" to onlinePlayer) {
                    then("period" to periodSuggest) {
                        then("unit" to unitSuggest) {
                            then("reason" to reasonString) {
                                executes {
                                    try {
                                        val target: String by it
                                        val reason: String by it
                                        val period: Long by it
                                        val unit: String by it
                                        val id: String =
                                            (LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMddhhmmssSS"))
                                                .toLong() * 5).toString()

                                        var moder: String

                                        if(isConsole) moder = "Server"
                                        else moder = (sender as Player).uniqueId.toString()

                                        val targetPlayer = Bukkit.getOfflinePlayer(target)

                                        if (!targetPlayer.hasPlayedBefore()) throw IllegalArgumentException("Targeted player not found")

                                        val managedPlayer = ManagedPlayer(
                                            Bukkit.getOfflinePlayer(target),
                                            PlayerManager.playerJSONFile
                                        )

                                        var toban = LocalDateTime.now()

                                        when(unit) {
                                            "days" -> toban = toban.plusDays(period)
                                            "hours" -> toban = toban.plusHours(period)
                                            "minutes" -> toban = toban.plusMinutes(period)
                                            "seconds" -> toban = toban.plusSeconds(period)
                                            else -> throw IllegalArgumentException("Unit Mismatched")
                                        }

                                        managedPlayer.tempban(reason, id, toban)

                                        if (managedPlayer.p.isOnline) {
                                            val onp: Player = Bukkit.getPlayer(managedPlayer.p.uniqueId)!!
                                            onp.showTitle(
                                                Title.title(
                                                    text("당신은 이 서버에서 차단되었습니다")
                                                        .color(color(0x00ff5555)),
                                                    text(managedPlayer.banre)
                                                        .color(color(0x00ffaa00))
                                                )
                                            )

                                            val kick: () -> Unit = {
                                                onp.kick(
                                                    text("당신은 이 서버에서 차단되었습니다\n\n${managedPlayer.getNokori(LocalDateTime.now(), "d일 h시간 m분 s초 남음")}\n\n").color(color(0x00ff5555)).append(
                                                        text("Reason: ").color(color(0x00aaaaaa))).append(
                                                        text(managedPlayer.banre).color(color(0x00ffffff))).append(
                                                        text("\n\nBan ID: ").color(color(0x00aaaaaa))).append(
                                                        text(managedPlayer.banid).color(color(0x00ffffff)))
                                                )
                                            }

                                            Bukkit.getScheduler().runTaskLater(
                                                io.github.kuroka3.playermanager.PlayerManager.instance,
                                                kick,
                                                100L
                                            )
                                        }

                                        sender.sendMessage(
                                            mod.append(
                                                text("${managedPlayer.p.name}").color(color(0x00ffff55))
                                            ).append(
                                                text("님을 서버에서 차단했습니다: ").color(color(0x00ff5555))
                                            ).append(
                                                text(reason).color(color(0x00ffaa00))
                                            )
                                        )

                                        BanIDManager.setBan(id, managedPlayer.p.uniqueId, moder, reason, LocalDateTime.now(), true, "$period$unit")
                                        CaseManager.addCase(4, moder, managedPlayer.p.uniqueId, reason, LocalDateTime.now(), "$period$unit")
                                    } catch (e: Exception) {
                                        when (e) {
                                            is IllegalArgumentException, is IllegalAccessException -> sender.sendMessage(
                                                mod.append(
                                                    text("${e.message}").color(color(0x00ff55555))
                                                )
                                            )

                                            else -> {
                                                e.printStackTrace()
                                                sender.sendMessage(
                                                    mod.append(
                                                        text("Check Console: An Exception occured")
                                                            .color(color(0x00ff55555))
                                                    )
                                                )
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}